core/lib/zksync_core/src/api_server/web3/pubsub.rs

```diff
@@ -78,18 +93,13 @@ impl PubSubNotifier<typed::Sink<PubSubResult>> {
             if let Some(last_block) = new_blocks.last() {
                 last_block_number = MiniblockNumber(last_block.number.unwrap().as_u32());
 
-                let notify_latency =
-                    PUB_SUB_METRICS.notify_subscribers_latency[&SubscriptionType::Blocks].start();
-                for sink in self.current_subscribers().await {
-                    for block in new_blocks.iter().cloned() {
-                        if sink.notify(Ok(PubSubResult::Header(block))).is_err() {
-                            // Subscriber disconnected.
-                            break;
-                        }
-                        PUB_SUB_METRICS.notify[&SubscriptionType::Blocks].inc();
+                for new_block in new_blocks {
+                    if !self.sender.receiver_count().is_zero() {
```

Comment:
This seems prone to TOCTOU issues: broadcast receivers may appear or disappear during the call. IMO, it would be more straightforward just to remove this condition.

Sort of related: One thing that makes `broadcast` channel different from the previous implementation is that a receiver can now receive *a part* of changes produced during a single `PubSubNotifier` iteration. I don't think it could be avoided, but it seems benign (although logs may require some thinking – is it OK if a subscriber receives logs starting from a middle of a certain miniblock?).

Reply:
removed.

Reply:
> s it OK if a subscriber receives logs starting from a middle of a certain miniblock?
No it's not ok

Reply:
@slowli wdyt about sending atomically a `Vec<PubSubEvent>` here? That would make it so that no matter what the client consistently sees _all_ things that have happend in a given block.

Reply:
Yep, that seems the most reasonable option.

Reply:
Done